modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipServiceTest.java

```diff
@@ -243,6 +271,111 @@ private void _testAddObjectRelationship(User user) throws Exception {
+			ServiceContextTestUtil.getServiceContext());
+
+		Role role = RoleTestUtil.addRole(RoleConstants.TYPE_REGULAR);
+
+		_resourcePermissionLocalService.setResourcePermissions(
+			TestPropsValues.getCompanyId(), objectDefinition2.getClassName(),
+			ResourceConstants.SCOPE_COMPANY,
+			String.valueOf(TestPropsValues.getCompanyId()), role.getRoleId(),
+			new String[] {ActionKeys.VIEW});
+
+		User user = UserTestUtil.addUser();
+
+		_userLocalService.addRoleUser(role.getRoleId(), user);
+
+		_setUser(user);
+
+		AssertUtils.assertFailure(
+			PrincipalException.MustHavePermission.class,
+			StringBundler.concat(
+				"User ", user.getUserId(), " must have UPDATE permission for ",
+				objectDefinition2.getClassName(), StringPool.SPACE,
+				objectEntry2.getObjectEntryId()),
+			() ->
+				_objectRelationshipService.
+					addObjectRelationshipMappingTableValues(
+						objectRelationship.getObjectRelationshipId(),
+						objectEntry1.getObjectEntryId(),
+						objectEntry2.getObjectEntryId(), new ServiceContext()));
+
+		_resourcePermissionLocalService.setResourcePermissions(
+			TestPropsValues.getCompanyId(), objectDefinition2.getClassName(),
+			ResourceConstants.SCOPE_COMPANY,
+			String.valueOf(TestPropsValues.getCompanyId()), role.getRoleId(),
+			new String[] {ActionKeys.UPDATE, ActionKeys.VIEW});
+
+		if (Objects.equals(
+				ObjectRelationshipConstants.TYPE_ONE_TO_MANY,
+				objectRelationshipType)) {
+
+			_objectRelationshipService.addObjectRelationshipMappingTableValues(
+				objectRelationship.getObjectRelationshipId(),
+				objectEntry1.getObjectEntryId(),
+				objectEntry2.getObjectEntryId(), new ServiceContext());
+
+			return;
+		}
+
+		_resourcePermissionLocalService.setResourcePermissions(
+			TestPropsValues.getCompanyId(), objectDefinition1.getClassName(),
+			ResourceConstants.SCOPE_COMPANY,
+			String.valueOf(TestPropsValues.getCompanyId()), role.getRoleId(),
+			new String[] {ActionKeys.VIEW});
+
+		AssertUtils.assertFailure(
+			PrincipalException.MustHavePermission.class,
+			StringBundler.concat(
+				"User ", user.getUserId(), " must have UPDATE permission for ",
+				objectDefinition1.getClassName(), StringPool.SPACE,
+				objectEntry1.getObjectEntryId()),
+			() ->
+				_objectRelationshipService.
+					addObjectRelationshipMappingTableValues(
+						objectRelationship.getObjectRelationshipId(),
+						objectEntry1.getObjectEntryId(),
+						objectEntry2.getObjectEntryId(), new ServiceContext()));
+
+		_resourcePermissionLocalService.setResourcePermissions(
+			TestPropsValues.getCompanyId(), objectDefinition1.getClassName(),
+			ResourceConstants.SCOPE_COMPANY,
+			String.valueOf(TestPropsValues.getCompanyId()), role.getRoleId(),
+			new String[] {ActionKeys.UPDATE, ActionKeys.VIEW});
+
+		_objectRelationshipService.addObjectRelationshipMappingTableValues(
+			objectRelationship.getObjectRelationshipId(),
+			objectEntry1.getObjectEntryId(), objectEntry2.getObjectEntryId(),
+			new ServiceContext());
```

Comment:
Before testing the success scenario (user with view and update permission), can you add the failure scenario where the user only has the update permission (the test should fail because of the missing permission of view)?

Reply:
Done [here](https://github.com/liferay-bpm/liferay-portal/pull/4167/commits/52a2dd3c9ee67fe8ff78cc993f34c737e7c6836d)